Initial implementation of jmeter websocket plugin

huzaus opened the pull request.

huzaus committed:
add initial git ignore

huzaus committed:
Fix copySample task

huzaus committed:
Implement initial UI

huzaus committed:
Implement openSampleMac gradle task

huzaus committed:
Rename WebsocketSampler -> WebsocketMessageSampler

huzaus committed:
Force copySample task execution

huzaus committed:
Implement tests for GUI class

huzaus committed:
Add JmeterAbstractSpec to address Jmeter setup

huzaus committed:
Implement WebsocketMessageSamplerModifier responsible for moving the data from your gui elements to the sampler

huzaus committed:
Implement WebsocketMessageSampler storage mechanism

huzaus committed:
Update sample test plan with predefined values

huzaus committed:
Implement WebsocketMessageSamplerConfigurer responsible for moving the data from the sampler to the gui elements

huzaus committed:
Update sample test plan with predefined values for ws and wss protocols

huzaus committed:
Remove redundant super() call

huzaus committed:
Remove instanceof check because only one implementation of TestElement should be supported

huzaus committed:
Rearrange custom tasks

huzaus committed:
Add CookieManager and HeaderManager support

huzaus committed:
Add SslContextFactory and URI

huzaus committed:
Add headers and upgradeRequest

huzaus committed:
Add cookies

huzaus committed:
Fix sample.jmx with correct path and message

huzaus committed:
Add explicit guava dependency

huzaus committed:
Implement WebsocketSessionManager which is responsible for storing web socket session

huzaus committed:
Add WebsocketEndpoint and WebsocketUpgradeListener dummy implementations

huzaus committed:
Add WebsocketMessageSampler dirty implementation

huzaus committed:
Refactor WebsocketSessionsManager toString method with unit tests

huzaus committed:
Refactor WebsocketEndpoint logging

huzaus committed:
Refactor WebsocketUpgradeListener logging

huzaus committed:
Add WebsocketEndpoint to WebsocketSessionManager

huzaus committed:
Implement onHandshakeRequest method

huzaus committed:
Implement onHandshakeResponse method

huzaus committed:
Improve WebsocketMessageSampler implementation

huzaus committed:
Rename Message -> Session in order to reflect responsibility

huzaus committed:
Rename package JMeter.plugins.functional.samplers -> com.jmeter.websocket.plugin

huzaus committed:
Add dummy WebsocketMessageSampler with gui

huzaus committed:
Move setting/getting WebsocketSessionsManager logic to AbstractWebsocketSampler

huzaus committed:
Implement dirty sample method

huzaus committed:
Change log levels and add temporary message text highlighted output

huzaus committed:
Remove future to simplify logic

huzaus committed:
Implement gui for WebsocketDataWriter

huzaus committed:
Move data writer logic to WebsocketSessionManager and delete WebsocketDataWriter

huzaus committed:
Move WebsocketEndpoint and WebsocketUpgradeListener to endpoint package

huzaus committed:
Encapsulate websocket client logic in WebsocketSession interface and add JettyWebsocketSession implementation

huzaus committed:
Extend WebsocketSession interface

huzaus committed:
Improve JettyWebsocketEndpoint with non-blocking writing to file

huzaus committed:
Refactor WebsocketSession

huzaus committed:
Improve logging for JettyWebsocketUpgradeListener

huzaus committed:
Fix lazy getter of WebsocketSession

huzaus committed:
Fix lazy getter of byteChannel

huzaus committed:
Improve write format

huzaus committed:
Replace SampleResult with HTTPSampleResult for better reporting

huzaus committed:
Implement toString method for JettyWebsocketEndpoint

huzaus committed:
Rename WebsocketSession -> WebsocketClient and change responsibility of the interface

huzaus committed:
Refactor massively in order to use mapping URI to Session

huzaus committed:
Implement tests for JettyWebsocketEndpoint

huzaus committed:
Implement tests for JettyWebsocketUpgradeListener

huzaus committed:
Implement tests for JettyWebsocket

huzaus committed:
Implement tests for stop and start methods of JettyWebsocketEndpoint

huzaus committed:
Add not null check for executor

huzaus committed:
Add check for sampleResult

huzaus committed:
Implement tests for CsvFileWriter

adgon92 commented on src/main/java/com/jmeter/websocket/plugin/endpoint/jetty/JettyWebsocket.java:
```diff
@@ -0,0 +1,76 @@
+
+    private static final Logger log = LoggingManager.getLoggerForClass();
+
+    private Collection<WebsocketMessageProcessor> websocketMessageProcessors = new ArrayList<>();
```
For the purpose of soft immutability, this field could be final.

adgon92 commented on src/main/java/com/jmeter/websocket/plugin/endpoint/jetty/JettyWebsocketEndpoint.java:
```diff
@@ -0,0 +1,158 @@
+    private Function<SampleResult, UpgradeListener> sampleResultToUpgradeListenerConverter = new SampleResultToUpgradeListenerConverter();
+
+    private Collection<WebsocketMessageProcessor> websocketMessageProcessors = new ArrayList<>();
+    private ConcurrentHashMap<URI, Session> sessions = new ConcurrentHashMap<>();
```
Whole class looks a bit messy. What I'd do is to move session management to a separate object. Enpoint should depend on session manager responses. E. g. there are a few methods where you check that session is not null nor opened.

adgon92 commented on src/main/java/com/jmeter/websocket/plugin/configurations/WebsocketSessionsManager.java:
```diff
@@ -0,0 +1,80 @@
+    private static final Logger log = LoggingManager.getLoggerForClass();
+
+    private static Supplier<CsvFileWriter> csvFileWriterSupplier;
+    private static Supplier<WebsocketClient> websocketClientSupplier = websocketClientSupplier();
```
Here i doubt that these values should be static. Are these objects really heavy weight or you use then like a "Singleton" objects??

huzaus replied:
@adgon92 We can have multiple instances of WebsocketSessionsManager (one instance per User) but we will use only one connection(WebsocketClient) and we will dump messages from all users to one file. So, yes we need single instance.

aplotnikov replied:
I would add final modifier.

huzaus committed:
Fix name JettyWebsocketEndpointSpock -> JettyWebsocketEndpointSpec

huzaus committed:
Fix code remark

huzaus committed:
Fix bug with unregisterWebsocketMessageConsumer((csvFileWriterSupplier.get()) on testEnded

See WebsocketSessionsManager :
    @Override
    public void testEnded() {
        getWebsocketClient().stop();
        csvFileWriterSupplier.get().stop();
        getWebsocketClient().unregisterWebsocketMessageConsumer(csvFileWriterSupplier.get());
        log.info("Test ended: " + this);
    }

aplotnikov commented on build.gradle:
```diff
@@ -11,20 +11,31 @@ repositories {
 
 dependencies {
-
     compile 'org.codehaus.groovy:groovy-all:2.4.7'
```
I would recommend update version of groovy to 2.4.8

huzaus replied:
👍

aplotnikov commented on .gitignore:
```diff
@@ -0,0 +1,3 @@
+.gradle/
```
I would update this file with bigger combination of exclusions e.g. exclusion for zip, jar, log, temp files etc. You can find many examples of such configuration in the internet or if you need I can propose one.

huzaus replied:
What is the point to add additional configuration? I like current gitignore setup because it is minimal, simple and cover all untracked files.

huzaus replied:
#3

aplotnikov commented on build.gradle:
```diff
@@ -11,20 +11,31 @@ repositories {
     compile 'org.apache.jmeter:ApacheJMeter:3.1'
     compile 'org.apache.jmeter:ApacheJMeter_core:3.1'
     compile 'org.apache.jmeter:ApacheJMeter_http:3.1'
+    compile 'com.google.guava:guava:19.0'
```
I believe guava 21.0 version was released.

huzaus replied:
Moreover I need to add it to Jmeter classpath

huzaus replied:
Issue to address this remark: #2

huzaus committed:
Fix code remark

Move session management to JettySessionManager

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/configurations/WebsocketSessionsManager.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    private static Supplier<WebsocketClient> websocketClientSupplier() {
+        return memoize(new Supplier<WebsocketClient>() {
```
Do you use Java 8?

huzaus replied:
Jmeter is running on java7 and above, so I can not use java8.

aplotnikov replied:
It is a pity (((

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/configurations/WebsocketSessionsManager.java:
```diff
@@ -0,0 +1,80 @@
+    private static Supplier<CsvFileWriter> csvFileWriterSupplier;
+    private static Supplier<WebsocketClient> websocketClientSupplier = websocketClientSupplier();
+
+    public String getFile() {
```
Maybe it should be filePath or something like this because when I am reading this line I am thinking about jdk File class

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/configurations/WebsocketSessionsManager.java:
```diff
@@ -0,0 +1,80 @@
+public class WebsocketSessionsManager extends ConfigTestElement implements TestStateListener {
+
+    private static final String FILE = "websocket.data.output.file";
+    private static final Logger log = LoggingManager.getLoggerForClass();
```
Are you able to use lombok?

huzaus replied:
Yep, but it will require Jmeter classpath update. Possible improvement for the next pull request.

huzaus replied:
#5

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/configurations/WebsocketSessionsManager.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Override
+    public void testStarted() {
+        csvFileWriterSupplier = csvFileWriterSupplier(Paths.get(getFile()));
```
maybe it is better have name of method like csvFileWriterSupplierOf(Paths.get(getFile()))

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/configurations/WebsocketSessionsManager.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    @Override
+    public void testStarted() {
```
I think it is better to change name of this method because how it looks in the usage
manager.testStarted I think it is better to have manager.startTests or so.

What do you think?

huzaus replied:
I can not change org.apache.jmeter.testelement.TestStateListener interface.

aplotnikov replied:
I see

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/configurations/WebsocketSessionsManager.java:
```diff
@@ -0,0 +1,80 @@
+import org.apache.jmeter.config.ConfigTestElement;
+import org.apache.jmeter.testelement.TestStateListener;
+import org.apache.jorphan.logging.LoggingManager;
+import org.apache.log.Logger;
```
I think you should improve logger because it uses apache implememntation. I would recommend to use slf4j API and use the same implementation which you have now but you won't have dependency in the code to any specific implementation and in case you need to change logger because of performance issue or so it will be easy to do.

huzaus replied:
Agree, I'll do this in future pull requests.

huzaus replied:
#4

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/configurations/WebsocketSessionsManager.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Override
+    public void testStarted(String host) {
+        log.info("Test started: " + this + ". Host: " + host + ".");
```
if you use slf4j logger you will be able to do something like 
`
log.info("Test started: {}. Host:{}.", this, host)
`

huzaus replied:
I know :)

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/configurations/WebsocketSessionsManagerGui.java:
```diff
@@ -0,0 +1,62 @@
+public class WebsocketSessionsManagerGui extends AbstractConfigGui {
+
+    private final FilePanel filePanel;
+    private static final String[] EXTENSIONS = {".csv"};
```
I believe code convention says that private static final fields (constants) have to be higher in the order than private final

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/configurers/Configurer.java:
```diff
@@ -0,0 +1,8 @@
+import org.apache.jmeter.samplers.AbstractSampler;
+import org.apache.jmeter.samplers.gui.AbstractSamplerGui;
+
+public interface Configurer<S extends AbstractSampler, T extends AbstractSamplerGui> {
```
Maybe it would be better to change T -> G like Gui in such case you will have S - sampler and G - Gui

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/endpoint/WebsocketClient.java:
```diff
@@ -0,0 +1,25 @@
+
+    void sendMessage(URI uri,String message) throws IOException;
+
+    void registerWebsocketMessageConsumer(WebsocketMessageProcessor processor);
```
I would improve name of this method. First of all websocket word is presented in name of class so I would remove at least this word from name of method. Also I would think it make it simple without message just registerConsumer. Another point here you register consumers but argument class is processor. I believe you have to have the same name for both cases.

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/endpoint/WebsocketClient.java:
```diff
@@ -0,0 +1,25 @@
+
+public interface WebsocketClient {
+
+    void connect(URI uri, CookieManager cookieManager, Map<String, List<String>> headers, SampleResult result, long timeOut) throws Exception;
```
Maybe it is better to use some class like Term or so for timeout you will be able to configure any timeout which you need like second, minute etc and it will be more readable. might be such class already exists in some library. However it is minor if it is used in one place only or so.

huzaus replied:
I prefer simpler solution for now as it is done in jmeter http request sampler: 
![image](https://cloud.githubusercontent.com/assets/17699437/22366110/32bbb47e-e47b-11e6-93e4-49dae4c059cb.png)

aplotnikov replied:
OK. Up to you.

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/endpoint/comsumers/CsvFileWriter.java:
```diff
@@ -0,0 +1,119 @@
+                                    action,
+                                    message.getBytes().length,
+                                    message
+
```
empty line

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/endpoint/comsumers/CsvFileWriter.java:
```diff
@@ -0,0 +1,119 @@
+                log.error("Exception thrown on write to byte channel: " + e);
+            }
+        } else {
+            log.info(message);
```
It looks like when we don't have bytechannel we have info log. Is it suitable? I suppose it is not normal case.

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/endpoint/comsumers/CsvFileWriter.java:
```diff
@@ -0,0 +1,119 @@
+            } catch (IOException e) {
+                log.error("Exception thrown on write to byte channel: " + e);
+            }
+        } else {
```
I would avoid such else case and it will remove 1 level of margin.

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/endpoint/comsumers/CsvFileWriter.java:
```diff
@@ -0,0 +1,119 @@
+            try {
+                byteChannel.close();
+            } catch (IOException e) {
+                log.error("Exception thrown on close byteChannel: " + e);
```
This log does not contain stack trace I would add too.

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/endpoint/comsumers/CsvFileWriter.java:
```diff
@@ -0,0 +1,119 @@
+                    return newByteChannel(file, OPEN_OPTIONS);
+                } catch (IOException e) {
+                    log.error("Exception thrown on open byte channel: " + e);
+                    return null;
```
It is a bad practise to return something from catch block because it is possible that information will lost from catch return because someone return something from finally block so I would move it outside of catch block.

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/endpoint/comsumers/WebsocketMessageProcessor.java:
```diff
@@ -0,0 +1,9 @@
+
+public interface WebsocketMessageProcessor {
+
+    void onMessageReceive(String sessionId, String message);
```
Maybe it makes sense to remove Message word here.

huzaus replied:
I would leave it for now.

aplotnikov replied:
Ok

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/endpoint/jetty/JettySessionManager.java:
```diff
@@ -0,0 +1,60 @@
+
+    @Override
+    public void closeSessions() {
+        for (Session session : sessions.values()) {
```
you can iterate collection via guava class as in Java 8 or very similar approach.

huzaus replied:
Unfortunately, FluentIterable doesn't have method for consumer, so I can not call  session.close().

aplotnikov replied:
It is a pity(((

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/endpoint/jetty/JettyWebsocketUpgradeListener.java:
```diff
@@ -0,0 +1,111 @@
+    @Override
+    public void onHandshakeRequest(UpgradeRequest request) {
+        checkNotNull(request);
+        log.debug("onHandshakeRequest() request: " + toStringHelper(request)
```
I think here is better to add isDebugMode or so I don't remember name of that method because you are doing to string operation for request and it can contain a lot of information and it can a bit impact performance. 

What do you think?

huzaus replied:
It can be included in scope of Logging improvements.

aplotnikov replied:
ok

huzaus replied:
#4

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/endpoint/jetty/JettyWebsocketUpgradeListener.java:
```diff
@@ -0,0 +1,111 @@
+                                        .join(headers);
+                            }
+                        })
+                        .or("")
```
I believe guava contains constant EMPTY

huzaus replied:
I've not found this constant in Guava.

aplotnikov replied:
It is a pity(((

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/endpoint/jetty/converters/SampleResultToUpgradeListenerConverter.java:
```diff
@@ -0,0 +1,22 @@
+
+public class SampleResultToUpgradeListenerConverter implements Function<SampleResult, UpgradeListener> {
+    @Override
+    public UpgradeListener apply(SampleResult sampleResult) {
```
Why don't you use Optional?

aplotnikov commented on src/main/java/com/jmeter/websocket/plugin/samplers/WebsocketMessageSampler.java:
```diff
@@ -0,0 +1,46 @@
+        sampleResult.setSampleLabel(getName());
+        try {
+            WebsocketClient websocketClient = getWebsocketClient();
+            checkNotNull(websocketClient, "WebsocketSessionManager should be added to test plan");
```
I think not 'should' -> 'has to'

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/JmeterAbstractSpec.groovy:
```diff
@@ -0,0 +1,16 @@
+import static org.apache.jmeter.util.JMeterUtils.setJMeterHome
+
+abstract class JmeterAbstractSpec extends Specification {
+    def setupSpec() {
```
according to codenarc convention you should not use def

huzaus replied:
In [spock sources](https://github.com/spockframework/spock-example) they are using def.

aplotnikov replied:
I would not rely on Spock source code. That code might be done many years ago before such best practice rule was introduced.

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/JmeterAbstractSpec.groovy:
```diff
@@ -0,0 +1,16 @@
+    def setupSpec() {
+        String jmeterHome = System.properties.'jmeter.home'
+        setJMeterHome(jmeterHome)
+        loadJMeterProperties(jmeterHome + '/libexec/bin/jmeter.properties')
```
I think it is better to use GString here

huzaus committed:
Update sample test plan to reflect latest changes

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/WebsocketUpgradeListenerSpock.groovy:
```diff
@@ -0,0 +1,91 @@
+    @Subject
+    JettyWebsocketUpgradeListener listener = new JettyWebsocketUpgradeListener(new SampleResult())
+
+    def "Should throw NullPointerException when result is null"() {
```
I would recommend to use void instead of def and String instead of GString.

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/WebsocketUpgradeListenerSpock.groovy:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    def "Should extract #headers upgrade request headers to sample result request headers as '#result'"() {
+        given:
```
I would recommend to use additional tabs for inner element under label (e.g. given...)
I mean something like this
````
given:
    UpgradeRequest upgradeRequest = new UpgradeRequest('wss://127.0.0.1/websocket')
when:
    listener.onHandshakeRequest(upgradeRequest)
````
From my point of view it is more readable.

huzaus replied:
@aplotnikov I agree. It is required to setup IDEA formatting and reformat all tests. Do you know how to do this?

aplotnikov replied:
Yes I know.
Codestyle -? Groovy -> tabs and indents -> label indent style

huzaus replied:
👍

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/WebsocketUpgradeListenerSpock.groovy:
```diff
@@ -0,0 +1,91 @@
+
+    def "Should extract '#statusReason' statusReason, '#responseCode' response code, #success response status and set it sample result"() {
+        given:
+        UpgradeResponse upgradeResponse = new UpgradeResponse()
```
I think you can create it something like the following:
`
new UpgradeResponse(statusCode: responseCode, ...)
`

huzaus replied:
👍

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/WebsocketUpgradeListenerSpock.groovy:
```diff
@@ -0,0 +1,91 @@
+        success == listener.sampleResult.successful
+        where:
+        statusReason     | responseCode | success
+        ''               | 101          | true
```
As I remember there are some class with constant where you have all this response code.
I believe if it is possible to use it here I would use it
However I don't remember maybe it is part os Spring and you don't have it here.

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/configurations/WebsocketSessionsManagerSpec.groovy:
```diff
@@ -0,0 +1,31 @@
+
+    def "Should return WebsocketClient with WebsocketMessageProcessor"() {
+        given:
+        manager.setFile(Files.createTempFile("temp-file", ".tmp").toString())
```
I would recommend to use junit rule which provides the same functionality but clean resources by the himself when test is finished.

huzaus replied:
#6

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/endpoint/comsumers/CsvFileWriterSpec.groovy:
```diff
@@ -0,0 +1,133 @@
+        when:
+        csvWriter.onMessageReceive('101', 'Message')
+        then:
+        1 * byteChannel.write(_) >> { throw new IOException() }
```
I am a bit confused with this check.
Could you explain me what you want to do here?
I feel something wrong here.

huzaus replied:
I want to check that exception will be caught in try catch block and won't be propagated further, because it can crash jmeter.

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/endpoint/jetty/JettySessionManagerSpec.groovy:
```diff
@@ -0,0 +1,95 @@
+    def "Should return '#value' if '#registeredUri' session is registered and it isOpen()='#isOpen' on hasOpenSession('#uri')"() {
+        given:
+        manager.registerSession(registeredUri,
+                Stub(Session) {
```
It looks like you use this stub in many test maybe it is better to make it as shared field or field. Be honest creating such instance looks bad here (((

huzaus replied:
It is not possible, because behaviour is parametrized by where block.

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/endpoint/jetty/JettySessionManagerSpec.groovy:
```diff
@@ -0,0 +1,95 @@
+        and:
+        1 * session.close()
+        and:
+        manager.sessions.isEmpty()
```
If you have mock maybe it makes sense to add such check at the end 
0 * _

huzaus replied:
It is questionable, I need to think about it.

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/endpoint/jetty/converters/CookieManagerToWebSocketClientConverterSpec.groovy:
```diff
@@ -0,0 +1,40 @@
+        when:
+        WebSocketClient client = converter.apply(null)
+        then:
+        client.executor != null
```
I would use with keyword. I believe it is perfect place for that word to use.

huzaus replied:
👍

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/endpoint/jetty/converters/HeadersToClientUpgradeRequestConverterSpec.groovy:
```diff
@@ -0,0 +1,34 @@
+
+    def "Should create new upgrade request with provided [#name:#values] headers on apply"() {
+        given:
+        def headers = [(name): values]
```
I would avoid using def especially in the code types.

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/endpoint/jetty/converters/SampleResultToUpgradeListenerConverterSpec.groovy:
```diff
@@ -0,0 +1,30 @@
+        when:
+        UpgradeListener listener = converter.apply(sampleResult)
+        then:
+        listener != null
```
I would just keep listener in such case it will valid it by default to null or not.

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/samplers/WebsocketSessionSamplerSpec.groovy:
```diff
@@ -0,0 +1,125 @@
+        sampler."$property" == null
+        where:
+        property << ["cookieManager", "headerManager"]
+
```
empty line

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/samplers/WebsocketSessionSamplerSpec.groovy:
```diff
@@ -0,0 +1,125 @@
+        '127.0.0.1'    | '8080'     | 'ws'     | '/websocket' | 'ws://127.0.0.1:8080/websocket'
+    }
+
+
```
empty line

aplotnikov commented on src/test/groovy/com/jmeter/websocket/plugin/samplers/WebsocketSessionSamplerSpec.groovy:
```diff
@@ -0,0 +1,125 @@
+        and:
+        def headers = sampler.headers()
+        then:
+        [(name): [value]] == headers
```
I would change order of variable.

huzaus committed:
Fix Andrii's remarks

huzaus committed:
Reformat code especially spock tests

huzaus committed:
Make websocketClientSupplier final

huzaus committed:
Refactor packages structure

huzaus committed:
Rename project

huzaus committed:
Add session Id text field to Websocket Session Sampler panel

huzaus committed:
Add session Id text field to Websocket Message Sampler panel and remove redundant fields

huzaus committed:
Replace URI with String for managing sessions

huzaus committed:
Update sample to demonstrate multiple sessions

huzaus committed:
Merge pull request #7 from huzaus/add-session-id

Remove session dependency on URI
